### ISSUE TITLE
Decide and document XDG media type

### ISSUE THREAD

**Issue by Akselmo** [robotframework/robotframework#4746]
I've created an issue to XDG mimetypes repo about adding a mediatype for robot framework
https://gitlab.freedesktop.org/xdg/shared-mime-info/-/issues/198

I think having mediatype `text/robotframework` would be ok. This means any text editor would know it's a text file, and those which have syntax highlight etc. support for Robot files can enable that as well.

**Comment by pekkaklarck** [comment]
An official media type sounds good. I don't know much about them, but `text/robotframework` sounds good. In addition to agreeing on that, we need to document it. Probably best to add an appendix to the User Guide where we could also document our file extensions and the registered remote server port.

**Comment by pekkaklarck** [comment]
The aforementioned changes to the User Guide are enough considering RF 6.1. I leave this issue open until we get the media type registered, though, and after that documentation also needs to be updated with a link to the IANA database.

**Comment by pekkaklarck** [comment]
I spend a little time studying media types to get a better understanding on then.Here are some resources:
- [IANA media types](https://www.iana.org/assignments/media-types/media-types.xhtml). Contains a bit of generic info, links to more info, and a list of registered media types.
- [IANA media type application form](https://www.iana.org/form/media-types).
- Various RFCs linked from the above pages. They are _very_ detailed and cover also other MIME related topics than media types. I only took a quick look on some of them.
- [Media type Wikipedia page](https://en.wikipedia.org/wiki/Media_type). This explains the media type syntax in simpler and more condensed manner than the aforementioned RFCs.

Based on my quick study, there are some things we need to decide:

1. Should we register to the "standards tree" or to the "vendor tree"? The former contains types like `text/plain` and `image/png`, and `text/robotframework` would fit there nicely. The latter is for publicly available products and it could be argued that Robot Framework would belong there along `application/vnd.ms-excel` and `image/vnd.adobe.photoshop`. If we'd go that route, then we also needed to use `text/vnd.robotframework`. I believe the standards tree is better, but I don't know are there restrictions with that.

2. Do we need to be able to separate suite files (`*.robot`) and resource files (`*.resource`)? If we do, how? The vendor tree would allow using a "tree prefix" like `texts/vnd.robotframework.resource`, but the standard tree doesn't. Media types also have a suffix, but it doesn't seem to be applicable in this context.

**Comment by Akselmo** [comment]
> Are you @Akselmo interested to look at the IANA registration? If you don't have time or interest, I can take care of that myself or ask could someone else do that. We can discuss more about that on Slack.

It's probably better if someone from RF foundation or anyone who works more closely with RF handles that. I have no idea how IANA works... Nor really have the time unfortunately.


> 2\. Do we need to be able to separate suite files (`*.robot`) and resource files (`*.resource`)? If we do, how? The vendor tree would allow using a "tree prefix" like `texts/vnd.robotframework.resource`, but the standard tree doesn't. Media types also have a suffix, but it doesn't seem to be applicable in this context.

If I understood it correctly, the XML file looks for file-ending, then for some "magic-string" within the file. I think those magic strings could be all the test data sections parts, such as `*** Settings ***`. We can be quite sure it's a RF `.robot` or `.resource` file if it has one of these sections within. 

Currently the files are thought to be patchfiles by many desktop environments that follow XDG standards, due to them having also the magic string of `*** stuff ...` However specifying that if the file has one of those asterisk sections *and* ends with a file-ending `.robot` or `.resources` should be enough for it to differentiate.

**Comment by pekkaklarck** [comment]
I don't think the person registering the media type needs to be that directly associated with the foundation. The [remote port](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=8270) was registered by someone from Oracle, but the foundation was the assignee. Totally understood if you don't have time or interest for this, though.

With the XML file do you mean how media types are configured with XDG, right? The [IANA application form](https://www.iana.org/form/media-types) also asks for things like file extensions and magic numbers. I believe magic numbers mean bytes in the beginning of the file and that isn't applicable with us.

The main open questions are:
- Should we use `text/robotframework`, as discussed earlier, or go with `text/vnd.robotframework`? I like the former more.
- Should we use same media type with both `.robot` and `.resource` files? I believe that's fine. Tools that want to show resource files separately can either look for the extension or does the file contain `*** Test Cases ***`.

Related to the latter point, I noticed that the example in the [issue you submitted](https://gitlab.freedesktop.org/xdg/shared-mime-info/-/issues/198) only contains `<glob pattern="*.robot"/>`. Shouldn't it also have `<glob pattern="*.resource"/>`?

**Comment by Akselmo** [comment]
> Totally understood if you don't have time or interest for this, though.

No time and not really interest either :sweat_smile: 

> * Should we use `text/robotframework`, as discussed earlier, or go with `text/vnd.robotframework`? I like the former more.

I am fine with either, former may be better. :)

>     * Should we use same media type with both `.robot` and `.resource` files? I believe that's fine. Tools that want to show resource files separately can either look for the extension or does the file contain `*** Test Cases ***`.

I think so yes, since they're both robot framework related items. No need for having multiple media types IMO.

> Related to the latter point, I noticed that the example in the [issue you submitted](https://gitlab.freedesktop.org/xdg/shared-mime-info/-/issues/198) only contains `<glob pattern="*.robot"/>`. Shouldn't it also have `<glob pattern="*.resource"/>`?

Yeah I seem to have forgot the resource file from that snippet :sweat_smile: It was also missing the magic number/strings.

**Comment by pekkaklarck** [comment]
I don't think [magic numbers](https://en.wikipedia.org/wiki/File_format#Magic_number) are applicable with us. Robot files often start with a header like `*** Settings ***`, but that's not guaranteed. First of all, any date before the first such header is just ignored and the header format also supports e.g. `* Settings`.

**Comment by pekkaklarck** [comment]
I looked at IANA media type registration again. `text/robotframework` would need to be registered in the standard tree and that process is pretty heavy. Getting it registered in the vendor tree would be easier, but then the media type would be `text/vnd.robotframework`. That would differ from what we decided to use with XDG which wouldn't be great. Probably best just use `text/robotframework` without official registration. It can be done later if there are concrete needs.
